src/middlewared/middlewared/plugins/ups.py

```diff
@@ -0,0 +1,157 @@
+import csv
+import io
+import os
+import re
+
+from middlewared.schema import accepts, Bool, Dict, Int, List, Str
+from middlewared.service import SystemServiceService, ValidationErrors
+from middlewared.validators import Email, ShouldBe
+
+
+class UPSService(SystemServiceService):
+
+    class Config:
+        datastore = 'services.ups'
+        datastore_prefix = 'ups_'
+        datastore_extend = 'ups.ups_config_extend'
+        service = 'ups'
+        service_verb = 'restart'
+
+    async def ups_config_extend(self, data):
+        data['mode'] = data['mode'].upper()
+        data['shutdown'] = data['shutdown'].upper()
+        data['toemail'] = [v for v in data['toemail'].split(';') if v]
+        return data
+
+    @accepts()
+    def ups_driver_choices(self):
+        ups_choices = {}
+        if os.path.exists("/conf/base/etc/local/nut/driver.list"):
+            with open('/conf/base/etc/local/nut/driver.list', 'rb') as f:
+                d = f.read().decode('utf8', 'ignore')
+            r = io.StringIO()
+            for line in re.sub(r'[ \t]+', ' ', d, flags=re.M).split('\n'):
+                r.write(line.strip() + '\n')
+            r.seek(0)
+            reader = csv.reader(r, delimiter=' ', quotechar='"')
+            for row in reader:
+                if len(row) == 0 or row[0].startswith('#'):
+                    continue
+                if row[-2] == '#':
+                    last = -3
+                else:
+                    last = -1
+                if row[last].find(' (experimental)') != -1:
+                    row[last] = row[last].replace(' (experimental)', '').strip()
+                for i, field in enumerate(list(row)):
+                    row[i] = field
+                ups_choices['$'.join([row[last], row[3]])] = '%s (%s)' % (' '.join(row[0:last]), row[last])
+        return ups_choices
+
+    async def validate_data(self, data, schema):
+        verrors = ValidationErrors()
+
+        driver = data.get('driver')
+        if driver:
+            if driver not in self.ups_driver_choices().keys():
+                verrors.add(
```

Comment:
`ups_driver_choices` does a lot of IO which may block the event loop. Its better if we call `self.middleware.call('ups.driver_choices')`